File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<vector<int>> board)
{
    int answer = 0;
    
    for(int i=0;i<board.size();i++){
        for(int j=0;j<board[i].size();j++){
            if(board[i][j]==1 && i>0 && j>0){
                board[i][j]=1+min({board[i-1][j], board[i][j-1], board[i-1][j-1]});
            }
        }
    }
    
    for(int i=0;i<board.size();i++){
        for(int j=0;j<board[i].size();j++){
            answer=max(answer,board[i][j]);
        }
    }

    return answer*answer;
}

File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.java
class Solution
{
    public int solution(int [][]board)
    {
        int answer = 1;
        int m = board.length;
        int n = board[0].length;
        int [][]nboard = new int[m+1][n+1];
        boolean check = true;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j] == 1){
                    check = false;
                    break;
                }
            }
        }
        if(check){
            return 0;
        }
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                nboard[i][j] = board[i][j];
            }
        }
        for(int i=1;i<m;i++){
            for(int j=1;j<n;j++){
                if(nboard[i][j] == 1){
                    nboard[i][j] = Math.min(nboard[i-1][j-1], Math.min(nboard[i-1][j], nboard[i][j-1])) + 1;
                }
                answer = Math.max(answer, nboard[i][j]);
            }
        }

        return answer * answer;
    }
}